proposed/swift-interop.md

```diff
@@ -130,7 +130,11 @@ CoreCLR and NativeAOT currently block the `VectorX<T>` types from P/Invokes as t
 
 ##### Automatic Reference Counting and Lifetime Management
 
-Swift has a strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. When .NET calls into Swift, the .NET GC is responsible for managing all managed objects. Unmanaged objects from C# should either implement `IDisposable` or utilize a designated thin wrapper over the Swift memory allocator, currently accessible through the `NativeMemory` class, to explicitly release memory. It's important to ensure that when a Swift callee function allocates an "unsafe" or "raw" pointer types, such as UnsafeMutablePointer and UnsafeRawPointer, where explicit control over memory is needed, and the pointer is returned to .NET, the memory is not dereferenced after the call returns. Also, if a C# managed object is allocated in a callee function and returned to Swift, the .NET GC will eventually collect it, but Swift will keep track using ARC, which represents an invalid case and should be handled by projection tools.
+Swift has a strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. When .NET calls into Swift, the .NET GC is responsible for managing all managed objects.
+
+The `IDisposable` provides an explicit mechanism for releasing unmanaged resources. Destructors are managed by the GC and offer a way to release unmanaged resources when an object is collected by the GC. While destructors abstract away memory management from the user, the `Idisposable` pattern provides deterministic control over when resources and can lead to better performance as it prevents the need for GC collection cycles. The `IDisposable` pattern is the typical .NET approach for dealing with unmanaged resources and thus is selected as default option at initial stage. If it is determined that the `IDisposable` pattern introduces unnecessary overhead for users, and that destructors can adequately manage the release of unmanaged resources, appropriate updates to the memory management approach will be made.
```

Comment:
> the Idisposable pattern provides deterministic control over when resources and can lead to better performance as it prevents the need for GC collection cycles

How is this relevant to interop? It sounds like it doesn’t make any semantic difference if the object is released early or late, it’s just a design difference. That is, it doesn’t sound like there are any non-memory resources being tracked. In that case I would expect us to stick with C# design, which is to use the GC for all memory resources. Is there something else being tracked here?

Reply:
> Is there something else being tracked here?

Composition can be a concern here. For example, in COM Aggregation there are potentially two distinct memory models that are represented by a single object - WinRT/COM class is the base of some .NET class. In the COM scenario this is relatively simple because it is expected the user to handle circular references that strattle the interop boundary. For WinRT, the Jupiter runtime has a reference tracking mechansim that means more cooperation is needed to collect a single object that is managed by the two systems.

I don't know if our Swift interop scenarios will have need for that sort of mechanism, but from experience interop scenarios that are non-deterministic make tooling very complicated - memory tracking, code coverage, etc. If we expect existing Swift tooling to be usable when .NET is involved having an explicit option is beneficial - even if it is an opt-in like [`CreateObjectFlags.UniqueInstance`].(https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.createobjectflags)

Reply:
The primary reason for considering destructors is to utilize the GC and the C# design pattern. The current tooling (BTfS) implements the `IDisposable` strategy. Since we are at an early stage, it is challenging to make a decision. 

> I don't know if our Swift interop scenarios will have need for that sort of mechanism, but from experience interop scenarios that are non-deterministic make tooling very complicated

Opting for `IDisposable` for the sake of simplicity seems like a reasonable argument. @agocke are you open to this approach (starting with `IDisposable`), or do you prefer starting with destructors initially and transitioning to `IDisposable` if required?

Reply:
My concern is mainly that we overload the use of `IDisposable` a lot in .NET, and that makes it confusing for customers. In some cases, like for files, it's pretty important that you call `IDisposable` to prevent running out of handles. In this case, pretty much every Swift type would come with `IDisposable`. Without knowing details of the type, it seems impossible for me, the caller, to understand whether I should or shouldn't dispose it.

Moreover, I think the preferred behavior for "trivial" deinit calls would be to **not** call Dispose (or more accurately, not wrap the type in a `using`). Since adding a `using` in C# produces a try-finally and hurts codegen, it's probably better to avoid a pattern that causes an explosion of usings. Instead, I think it would be ideal if we only used IDisposable for custom deinit. In those cases, we can be relatively confident that something more than just memory is being freed.

So overall I'm very supportive of some way to explicitly call deinit/dispose on Swift type, but making everything IDisposable feels like an anti-pattern that will just cause a lot of user confusion.